govuk/sign-in/token/index.js

```diff
@@ -1,30 +1,96 @@
+        if (!authorisationCode) {
+            throw Error('Authorisation code is required');
+        }
+
+        const jwtService = createJwtService();
+        const opts = {
+            url: `${process.env.CW_GOVUK_ISSUER_URL}/token`,
             headers: {
                 'Content-Type': 'application/x-www-form-urlencoded'
             },
             form: {
                 grant_type: 'authorization_code',
-                code: options.code,
-                redirect_uri: options.redirect_uri,
+                code: authorisationCode,
+                redirect_uri: getSignedInURI(),
                 client_assertion_type: 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer',
-                client_assertion: {/!* JWT token goes here *!/ }
+                client_assertion: jwtService.generateJWT({
+                    aud: `${process.env.CW_GOVUK_ISSUER_URL}/token`,
+                    iss: process.env.CW_GOVUK_CLIENT_ID,
+                    sub: process.env.CW_GOVUK_CLIENT_ID,
+                    exp: Date.now() + 1000 * 60 * 5, // 5 minutes from now.
+                    jti: uuidv4(),
+                    iat: Date.now()
+                })
             }
         };
 
-        return request.post('https://oidc.integration.account.gov.uk/token', opts); */
-        console.log(options);
+        const response = await post(opts);
+        return response.body;
+    }
+
+    async function validateIdToken(idToken) {
+        let decodedToken;
+        try {
+            const clientId = process.env.CW_GOVUK_CLIENT_ID;
+            const nonce = 'NONCE'; // TODO: ad to env vars.
+            const issuerService = createIssuerService();
+            const issuer = await issuerService.identify(); // TODO: pass in issuer in funtion signature.
+
+            const jwtService = createJwtService();
+            // remove milliseconds precision to match the token precision.
+            const now = parseInt(`${Date.now()}`.slice(0, -3), 10);
+
+            await jwtService.verifyJWT(idToken);
+
+            decodedToken = jwtService.decodeJWT(idToken);
+
+            // https://docs.sign-in.service.gov.uk/integrate-with-integration-environment/integrate-with-code-flow/#validate-your-id-token
+            if (decodedToken.iss !== issuer.issuer) {
+                throw Error(`issuer "${decodedToken.iss}" must be ${issuer.issuer}`);
+            }
+
+            if (decodedToken.aud !== clientId) {
+                throw Error('client ID mismatch!');
+            }
+
+            if (decodedToken.exp <= now) {
+                throw Error('token has expired!');
+            }
+
+            if (decodedToken.nonce !== nonce) {
+                throw Error('nonce mismatch!');
+            }
+
+            if (!(decodedToken.iat <= now && now <= decodedToken.exp)) {
+                throw Error('outwith token issued date and token expiry date!');
+            }
+        } catch (err) {
+            console.log({err});
+        }
```

Comment:
This should all be done by the JWT module. Some of it is out of your control, if the `exp` has expired for example, the `jsonwebtoken` verify function will fail regardless of any of these checks.